src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala

```diff
@@ -22,32 +22,26 @@ object BlockMatrix {
 
   def from(sc: SparkContext, lm: BDM[Double], blockSize: Int): M = {
     assertCompatibleLocalMatrix(lm)
-    val gp = GridPartitioner(blockSize, lm.rows, lm.cols)
+    val part = GridPartitioner(blockSize, lm.rows, lm.cols)
     val lmBc = sc.broadcast(lm)
-    val rowPartitions = gp.nBlockRows
-    val colPartitions = gp.nBlockCols
-    val indices = for {
-      i <- 0 until rowPartitions
-      j <- 0 until colPartitions
-    } yield (i, j)
-    val blocks = sc.parallelize(indices).map { case (i, j) =>
-      val iOffset = i * blockSize
-      val jOffset = j * blockSize
-      val rowsInBlock = gp.blockRowNRows(i)
-      val colsInBlock = gp.blockColNCols(j)
-      val a = new Array[Double](rowsInBlock * colsInBlock)
-      var jj = 0
-      while (jj < colsInBlock) {
-        var ii = 0
-        while (ii < rowsInBlock) {
-          a(jj * rowsInBlock + ii) = lmBc.value(iOffset + ii, jOffset + jj)
-          ii += 1
-        }
-        jj += 1
+    new BlockMatrix(
+    new RDD[((Int, Int), BDM[Double])](sc, Nil) {
+      override val partitioner = Some(part)
+
+      def getPartitions: Array[Partition] = Array.tabulate(part.numPartitions)(i => IntPartition(i))
+
+      def compute(split: Partition, context: TaskContext): Iterator[((Int, Int), BDM[Double])] = {
+        val (i, j) = part.blockCoordinates(split.index)
+        val iOffset = i * blockSize
+        val jOffset = j * blockSize
+        val (blockNRows, blockNCols) = part.blockDims(split.index)
+        // FIXME return slice when write supports sliced blocks
```

Comment:
Don't we always want to copy the sliced blocks to fresh BDMs? Otherwise every block of the BlockMatrix will carry around the full original array.

Reply:
What's wrong with that?

Reply:
I see, I should have a lazy model in mind. If I turn a 5GB matrix of eigenvectors of 25k by 25k LD into a block matrix and then write the latter to disk with support for slicing, then it's a win to not have to make the intermediate copy. I'm wondering if I have a two stage pipeline, with a persist on the first which makes the BlockMatrix, if the 5GB could be GCed during the second stage.

Spark doesn't support slicing on DenseMatrix; we have to decide if adding that complexity on blocks of BlockMatrix is worth it. Besides on read/write, we use `assertCompatibleLocalMatrix` in BlockMatrix to enforce offset=0 and the expected majorStride, which allows for using the data directly.

Reply:
I'd say if you don't want the whole matrix in memory you shouldn't broadcast the whole thing.  Broadcasting the the entire matrix only to persist a tiny portion is a terrible way to get the blocks persisted across the cluster.  This implementation should only be used for small matrices.